miden-crypto/src/merkle/smt/history/mod.rs

```diff
@@ -0,0 +1,429 @@
+    /// the method returns [`None`].
+    ///
+    /// # Gotchas
+    ///
+    /// When constructing the `nodes` and `leaves`, keep in mind that those collections must contain
+    /// entries for the **default value of a node or leaf** at any position where the tree was
+    /// sparse in the state represented by `root`. If this is not done, incorrect values may be
+    /// returned.
+    ///
+    /// This is necessary because the changes are the _reverse_ from what one might expect. Namely,
+    /// the changes in a given version `v` must "_revert_" the changes made in the transition from
+    /// version `v` to version `v + 1`.
+    ///
+    /// # Errors
+    ///
+    /// - [`HistoryError::NonMonotonicVersions`] if the provided version is not greater than the
+    ///   previously added version.
+    pub fn add_version(
+        &mut self,
+        root: Word,
+        version_id: VersionId,
+        nodes: NodeChanges,
+        leaves: LeafChanges,
+    ) -> Result<()> {
+        if let Some(v) = self.deltas.iter().last() {
+            if v.version_id < version_id {
+                self.deltas.push_back(Delta::new(root, version_id, nodes, leaves));
+                if self.num_versions() > self.max_versions() {
+                    self.deltas.pop_front();
+                }
+
+                Ok(())
+            } else {
+                Err(HistoryError::NonMonotonicVersions(version_id, v.version_id))
+            }
+        } else {
+            self.deltas.push_back(Delta::new(root, version_id, nodes, leaves));
+
+            Ok(())
+        }
+    }
+
+    /// Returns the index in the sequence of deltas of the version that corresponds to the provided
+    /// `version_id`.
+    ///
+    /// To "correspond" means that it either has the provided `version_id`, or is the newest version
+    /// with a `version_id` less than the provided id. In either case, it is the correct version to
+    /// be used to query the tree state in the provided `version_id`.
+    ///
+    /// # Complexity
+    ///
+    /// Finding the latest corresponding version in the history requires a linear traversal of the
+    /// history entries, and hence has complexity `O(n)` in the number of versions.
+    ///
+    /// # Errors
+    ///
+    /// - [`HistoryError::VersionTooOld`] if the history does not contain the data to provide a
+    ///   coherent overlay for the provided `version_id` due to `version_id` being older than the
+    ///   oldest version stored.
+    fn find_latest_corresponding_version(&self, version_id: VersionId) -> Result<usize> {
+        // If the version is older than the oldest, we error.
+        if let Some(oldest_version) = self.deltas.front() {
+            if oldest_version.version_id > version_id {
+                return Err(HistoryError::VersionTooOld);
+            }
+        } else {
+            return Err(HistoryError::VersionTooOld);
```

Comment:
I wonder if `VersionTooOld` is the right error here. For example, this could happen not because the version is too old, but because the tree just hasn't been updated recently. Not sure what's the best error here would be though.